library/alloc/src/boxed.rs

```diff
@@ -1016,6 +1016,177 @@ impl<T: ?Sized, A: Allocator> Box<T, A> {
         // additional requirements.
         unsafe { Pin::new_unchecked(boxed) }
     }
+
+    /// Attempts to reuse the storage allocated for this box to store a new value.
+    ///
+    /// # Errors
+    ///
+    /// This function fails if `T` and `U` do not have the same [memory layout](Layout). If you
+    /// instead want to create a new `Box` in this case, use [`Box::set`].
+    ///
+    /// # Examples
+    ///
+    /// Reuse a boxed slice:
+    ///
+    /// ```
+    /// #![feature(box_reuse)]
+    ///
+    /// let mut numbers: Box<[i32]> = Box::new([1, 2, 3, 4, 5]);
+    /// Box::reuse(&mut numbers, [6, 7, 8, 9, 10]).unwrap();
+    /// assert_eq!(&*numbers, &[6, 7, 8, 9, 10]);
+    /// ```
+    ///
+    /// Reuse a box to store an opaque type:
+    ///
+    /// ```
+    /// #![feature(box_reuse)]
+    ///
+    /// fn iterate() -> impl Iterator<Item = i32> {
+    ///     0..3
+    /// }
+    ///
+    /// let mut iterator: Box<dyn Iterator<Item = i32>> = Box::new(iterate());
+    /// assert_eq!(iterator.by_ref().collect::<Vec<_>>(), [0, 1, 2]);
+    ///
+    /// assert!(Box::reuse(&mut iterator, iterate()).is_ok());
+    /// assert_eq!(iterator.collect::<Vec<_>>(), [0, 1, 2]);
+    /// ```
+    #[unstable(feature = "box_reuse", issue = "none")]
+    pub fn reuse<U>(b: &mut Box<T, A>, val: U) -> Result<(), U>
+    where
+        U: Unsize<T>,
+    {
+        if Layout::for_value(&**b) != Layout::new::<U>() {
+            return Err(val);
+        }
+
+        struct Guard<'a, T: ?Sized, A: Allocator, U: Unsize<T>> {
+            b: &'a mut Box<T, A>,
```

Comment:
The pointee may (and will!) have been deallocated:

```suggestion
        struct Guard<'a, T: ?Sized, A: Allocator, U: Unsize<T>> {
            b: &'a mut Box<MaybeUninit<T>, A>,
```

Reply:
It's sound to hold a `Box` to a dropped value as long as that value is not accessed, right? Otherwise `ManuallyDrop::drop` would always be unsound.

Reply:
Yes, `ManuallyDrop<T>` could be used here too. It is unclear whether using none, as you do here, is a safety or a validity invariant, so I'd prefer to err on the side of caution.

Reply:
Would it be sound to transmute `&mut Box<T, A>` to `&mut Box<ManuallyDrop<T>, A>`? I know it would be sound for `&mut Box<T>`, but does that guarantee still apply when there is a non-ZST allocator? `Box` doesn't have an explicit `repr`.

Reply:
Ah, true. I'd say that it must, in order for `Box::from_raw(Box::into_raw(…).cast::<ThingWithSameLayout>(…))` or `Box::new_uninit_in().write(…)` to work. Maybe we should open an issue and ask on the UCG repo?

Reply:
Until that gets clarified, I guess the side of caution now leans on your side 😉